scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -323,127 +327,196 @@ public Pipeline getPipeline() {
+        Collectors.toMap(
+            (ResourceId dir) -> dir,
+            (ResourceId dir) -> new SMBFilenamePolicy(dir, filenameSuffix).forDestination()
+        )
+      );
       this.tupleTag = tupleTag;
-      this.inputDirectory = inputDirectory;
       this.filenameSuffix = filenameSuffix;
       this.fileOperations = fileOperations;
-
-      this.fileAssignment = new SMBFilenamePolicy(inputDirectory, filenameSuffix).forDestination();
     }
 
     private BucketedInput(
         TupleTag<V> tupleTag,
-        ResourceId inputDirectory,
+        List<ResourceId> inputDirectories,
         String filenameSuffix,
         FileOperations<V> fileOperations,
-        BucketMetadata<K, V> metadata) {
-      this(tupleTag, inputDirectory, filenameSuffix, fileOperations);
+        Map<ResourceId, BucketMetadata<K, V>> metadata) {
+      this(tupleTag, inputDirectories, filenameSuffix, fileOperations);
       this.metadata = metadata;
+      getCanonicalMetadata();
     }
 
     public TupleTag<V> getTupleTag() {
       return tupleTag;
     }
 
-    public BucketMetadata<K, V> getMetadata() {
-      if (metadata != null) {
-        return metadata;
-      } else {
-        try {
-          metadata =
-              BucketMetadata.from(
-                  Channels.newInputStream(FileSystems.open(fileAssignment.forMetadata())));
-          return metadata;
-        } catch (IOException e) {
-          throw new RuntimeException(e);
-        }
+    /**
+     Returns a single metadata to represent this source. Used for key extraction
+     functions and computing least # buckets (see: {@link MergeBuckets#leastNumBuckets})
+     */
+    public BucketMetadata<K, V> getCanonicalMetadata() {
+      if (canonicalMetadata == null) {
+        canonicalMetadata = getMetadata().values().stream()
+            .min(Comparator.comparingInt(BucketMetadata::getNumBuckets))
+            .get();
+      }
+
+      return canonicalMetadata;
+    }
+
+    public Map<ResourceId, BucketMetadata<K, V>> getMetadata() {
+      if (metadata == null) {
+        metadata = fileAssignments.entrySet().stream().collect(
+            Collectors.toMap(
+                Entry::getKey,
+                entry -> {
+                  try {
+                    return BucketMetadata.from(
+                        Channels.newInputStream(FileSystems.open(entry.getValue().forMetadata())));
+                  } catch (IOException e) {
+                    throw new RuntimeException(e);
+                  }
+                }));
       }
+
+      return metadata;
     }
 
     public Coder<V> getCoder() {
       return fileOperations.getCoder();
     }
 
-    private KeyGroupIterator<byte[], V> createIterator(int bucketId, int leastNumBuckets) {
-      // Create one iterator per shard
-      final BucketMetadata<K, V> metadata = getMetadata();
+    // Validates that all the inputDirectories in this source contain compatible metadata
+    void validateIntraSourceCompatibility() {
```

Comment:
This can be `validatePartitionCompatibility()` if we standardize on `partition` as the abstraction name?